### ISSUE TITLE
Altering - cluster-altering app shows error after upgrading Rancher

### ISSUE THREAD

**Issue by jiaqiluo** [rancher/rancher#21323]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
- run Rancher: v2.2.4 container
- as admin, enable Github auth
- log in as Github user user1, add an RKE DO cluster
- as user1, adds cluster notifier and cluster alerting
- upgrade Rancher to v2.2.5-rc8

**Result:**
- the cluster altering still works, but `cluster-altering` app shows the following error
<img width="1480" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/60628988-c2379580-9da8-11e9-983b-e9268412a9f6.png">

<img width="1424" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/60628645-3cffb100-9da7-11e9-9988-802475bee160.png">

- and it will cause the cluster unavailable periodicity
<img width="1412" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/60628652-438e2880-9da7-11e9-9ba7-4fdd44ed6603.png">

```
2019/07/03 22:03:05 [ERROR] Failed to send alert: Post http://alertmanager-operated.cattle-prometheus.svc.cluster.local:9093/api/v1/alerts: dial tcp: lookup alertmanager-operated.cattle-prometheus.svc.cluster.local on 67.207.67.2:53: no suc
2019/07/03 22:03:05 [ERROR] Failed to send alert: Post http://alertmanager-operated.cattle-prometheus.svc.cluster.local:9093/api/v1/alerts: dial tcp: lookup alertmanager-operated.cattle-prometheus.svc.cluster.local on 67.207.67.2:53: no suc
2019/07/03 22:03:05 [ERROR] Failed to send alert: Post http://alertmanager-operated.cattle-prometheus.svc.cluster.local:9093/api/v1/alerts: dial tcp: lookup alertmanager-operated.cattle-prometheus.svc.cluster.local on 67.207.67.2:53: no suc
```

- and after the API server recovers and the cluster is `active`, the app `cluster-alerting` is still in the bad state as the above screenshot shows.

**Comment by loganhz** [comment]
It looks like api server is unhealthy or it is due to network issue. All apps will hit this issue not just alerting.

**Comment by jiaqiluo** [comment]
@loganhz shows me that the `cluster-alerting` app get redeployed successfully and the error is gone after editing and save it.

**Comment by jiaqiluo** [comment]
Tried the same test in a new setup, and did not see the error happen again.
It is likely a timing issue, and the workaround is mentioned in the previous comment.

**Comment by loganhz** [comment]
After after upgrade, Rancher will upgrade the cluster alerting app as we introduced a new param for it.
However, Rancher cannot reach the K8s cluster. So the upgrade for the alerting app failed. Now you have to click edit and save to trigger the update manually.

As it’s not just for alerting app, users cannot upgrade apps if api server is not healthy. 

Maybe the correct fix should be adding some re-try logic in helm-controller.

**Comment by jiaqiluo** [comment]
The following comment is copied from https://github.com/rancher/rancher/issues/21364#issuecomment-509412324
 
Steps:
- run Rancher: v2.2.4 container
- as admin, enable Github auth
- log in as Github user user1
- add the following three clusters: DO single-node, DO 3-node, and GKE 
- adds cluster notifier and cluster alerting for all three clusters
- upgrade Rancher to v2.2.5-rc9

Results:
- in the clusters DO single-node and GKE, the app `cluster-alerting` and `monitoring-operator` get upgraded successfully, meaning there is only 1 workload left under the app `cluster-alerting`, and the altering feature is functional 
- in the cluster DO 3-node, the app `cluster-alerting` does NOT get upgraded, meaning there are still 2 workloads under the app  `cluster-alerting`, but the cluster altering still works
- and Rancher's log shows  
```
2019/07/08 20:56:00 [ERROR] ProjectController c-44pq6/p-jmpkk [system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready
2019/07/08 20:56:00 [ERROR] ProjectController c-44pq6/p-scrx7 [system-image-upgrade-controller] failed with : upgrade system service alerting failed, cluster c-44pq6 not ready
```


More Info 1
- adding a new single-node cluster and a new 3-node cluster in the same setup and enabling the cluster altering fails
- Rancher shows the following errors
```
Failed to ensure monitoring project name: failed to move Namespace cattle-prometheus into a Project: Operation cannot be fulfilled on namespaces "cattle-prometheus": the object has been modified; please apply your changes to the latest version and try again
```
- Rancher's log shows
 
```
2019/07/08 21:31:58 [ERROR] ClusterAlertGroupController c-28fpf/cag-8cwmg [cluster-alert-group-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-group-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
2019/07/08 21:32:00 [ERROR] ClusterAlertRuleController c-28fpf/car-8fw8v [cluster-alert-rule-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-rule-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
2019/07/08 21:32:04 [ERROR] ClusterAlertGroupController c-28fpf/cag-8cwmg [cluster-alert-group-deployer] failed with : failed to get Alertmanager Deployment information, statefulsets.apps "alertmanager-cluster-alerting" not found, [cluster-alert-group-controller] failed with : Failed to get service for alertmanager, service "cattle-prometheus/alertmanager-operated" not found
```


<img width="1408" alt="screenshot" src="https://user-images.githubusercontent.com/6218999/60845866-7ca50f00-a192-11e9-8dd4-09d8c849480a.png">

More Info 2
- tried the same steps in a fresh-installed Rancher v2.2.5-rc9 setup and the cluster altering works with no issues
- tried the same steps in a fresh-installed Rancher v2.2.5-rc8 setup and the cluster altering works with no issues

-------

Update 1: 
After about 2 hours, those two new clusters become healthy and the cluster altering starts working.
But for the cluster created before upgrading the app `cluster-alerting` still have 2 workloads.

**Comment by jiaqiluo** [comment]
The same tests as the above comment describe, but to v2.2.5-rc12
The results are positive 

Steps:
- run Rancher: v2.2.4 container
- as admin, enable Github auth
- log in as Github user user1
- add the following three clusters: DO single-node, DO 3-node, and GKE
- adds cluster notifier and cluster alerting for all three clusters
- upgrade Rancher to v2.2.5-rc12

Results:
- for all three clusters, the cluster alerting upgrade successfully
- the cluster altering still work
- adding a new alter group and new rule also works

One more test:
- add a new DO single-node cluster in the same setup 
- enable the cluster altering 

Results:
- in the beginning, Rancher logs show the following error only one time, and then the alerting start working
```
2019/07/10 00:32:00 [ERROR] Failed to send alert: Post http://alertmanager-operated.cattle-prometheus.svc.cluster.local:9093/api/v1/alerts: dial tcp: lookup alertmanager-operated.cattle-prometheus.svc.cluster.local on 10.43.0.10:53: no such
```

**Comment by jiaqiluo** [comment]
The same tests as the above comment describe, but from v2.2.3 to v2.2.5-rc12
The results are positive 

Steps:
- run Rancher: v2.2.3 container
- as admin, enable Github auth
- log in as Github user user1
- add the following three clusters: DO single-node, DO 3-node, and GKE
- adds cluster notifier and cluster alerting for all three clusters
- upgrade Rancher to v2.2.5-rc12

Results:
- for all three clusters, the cluster altering still work
- for all three clusters, adding a new alter group and new rule also works
- for all three clusters, the app `cluster-alerting` is stuck in `installing` state, like this:
<img width="1463" alt="from v2 2 3 to v2 2 5-rc12" src="https://user-images.githubusercontent.com/6218999/60932137-7f2a6600-a271-11e9-95f5-2754e7563f48.png">

- the app `cluster-alerting` in one cluster shows the following error
```
Failed to install app cluster-alerting. Error: UPGRADE FAILED: the server is currently unable to handle the request (get configmaps)
```
<img width="1479" alt="Screen Shot 2019-07-09 at 5 47 48 PM" src="https://user-images.githubusercontent.com/6218999/60932185-b13bc800-a271-11e9-9b6e-9bdc6c4f3c87.png">

- in another two clusters, it shows this error
```
Failed to install app cluster-alerting. Error: stat /rancher-monitoring: no such file or directory
```
<img width="1456" alt="from v2 2 3 to v2 2 5-rc12 " src="https://user-images.githubusercontent.com/6218999/60932219-d6c8d180-a271-11e9-96c9-4f6fc6a3fc97.png">


One more test:
- add a new DO single-node cluster in the same setup 
- enable the cluster altering 

Results:
- the cluster altering works

**Comment by jiaqiluo** [comment]
The bug fix is validated in the upgrade path from v2.2.3 to master-head

**Steps to reproduce (least amount of steps as possible):**
- run Rancher:v2.2.3 container 
- as admin, enable Github auth
- log in as Github user user1
- add 4 EC2 single-node clusters, and configure them:
  - only-alerting
  - only-alerting
  - only-logging
  - both-logging-alerting
- upgrade Rancher to master-head e65f73a9

Results:
- `cluster-alerting` and `rancher-logging` apps upgrade to the latest versions automatically after upgrading Rancher.

One more test:
- add a new DO single-node cluster in the same setup 
- enable the cluster altering 

Results:
- the cluster altering works

Known Issue:
- the `unable to handle the request error`  still exists and has been tracked here:
https://github.com/rancher/rancher/issues/21480
